### ISSUE TITLE
Error messages not displayed with crispy forms 1.7.2 - we had to downgrade to django-crispy-forms 1.6.1

### ISSUE THREAD

**Issue by uri-rodberg** [django-crispy-forms/django-crispy-forms#889]
* Package version: django-crispy-forms==1.7.2
* Django version: Django==1.11.23
* Python version: 3.5 & 3.6
* Template pack: 
```
CRISPY_TEMPLATE_PACK = 'bootstrap4'
CRISPY_FAIL_SILENTLY = False
```

### Description:

I need help. I don't know how to display error messages in crispy forms. Currently error messages which are not related to a specific field are not displayed. Error messages related to a specific field are displayed. I need to display all the error messages on the form and I don't know how? I didn't find it on the documentation of crispy forms.

You can see my current code on https://github.com/speedy-net/speedy-net/tree/staging
The forms not working right now with error messages are with URL /matches/settings/about-my-match/
The template only contains `{% crispy form %}`.

### Preferably also include:

- [ ] Example Django Crispy Forms code
- [ ] Screenshots

When I select a min age higher than max age I do see an error message:
![2019-08-07_1300_crispy_1](https://user-images.githubusercontent.com/12808124/62614231-ada06d00-b913-11e9-8b9e-6fb1491f71e3.png)

But when I don't select any gender there is no error message:
![2019-08-07_1300_crispy_2](https://user-images.githubusercontent.com/12808124/62614278-c27d0080-b913-11e9-8893-86cc6b450fa7.png)

Also, when the diet (for example) doesn't contain 5 hearts I can't see the error message:
![2019-08-07_1304_crispy_3](https://user-images.githubusercontent.com/12808124/62614465-1be52f80-b914-11e9-8322-e43d22fcc9bd.png)

Error messages for fields such as min age and max age are displayed correctly.
![2019-08-07_1307_crispy_4](https://user-images.githubusercontent.com/12808124/62614696-9910a480-b914-11e9-849a-fc6946eabe15.png)

In all the 4 cases above the user is not redirected to the success URL.

- [ ] Actual HTML generated
- [ ] Expected HTML

**Comment by carltongibson** [comment]
Hmmm. They should be shown: 

https://github.com/django-crispy-forms/django-crispy-forms/blob/7d7fc345986bf00b7047181448af71eaec28ebcb/crispy_forms/templates/uni_form/uni_form.html#L2-L4

Not sure why they wouldn't be... Can you check your context there?

**Comment by uri-rodberg** [comment]
@carltongibson The view used is `EditMatchSettingsView` and you can see the code here:

https://github.com/speedy-net/speedy-net/blob/staging/speedy/match/matches/views.py#L58

It uses a generic `UpdateView`.

We have a few forms inheriting from `SpeedyMatchProfileBaseForm` and this form (`SpeedyMatchProfileFullMatchForm`) is one of them.

How do I check the context?

**Comment by carltongibson** [comment]
Django Debug Toolbar is great for that. (It'll show you the exact context that was used to render the template.)

**Comment by uri-rodberg** [comment]
OK, here are the screenshots after running the form locally with 2 errors:

![2019-08-07_1325_crispy_local_5](https://user-images.githubusercontent.com/12808124/62615999-248b3500-b917-11e9-9f9f-3c1d3ffa3834.png)

![2019-08-07_1325_crispy_debug_5](https://user-images.githubusercontent.com/12808124/62616006-28b75280-b917-11e9-879b-01df5b4f1658.png)

**Comment by carltongibson** [comment]
But that's not the crispy template. You need to look for the base form template that crispy uses for your template pack (corresponding to the one I linked above.)

**Comment by uri-rodberg** [comment]
@carltongibson I have 30 crispy forms templates. Which one?

Here is the last one:

```
bootstrap4/layout/baseinput.html
...\\lib\\site-packages\\crispy_forms\\templates\\bootstrap4\\layout\\baseinput.html
Toggle context
{'False': False, 'None': None, 'True': True}
{'DEFAULT_MESSAGE_LEVELS': {'DEBUG': 10,
                            'ERROR': 40,
                            'INFO': 20,
                            'SUCCESS': 25,
                            'WARNING': 30},
 'False': False,
 'LANGUAGES': [('en', 'English'), ('he', 'Hebrew')],
 'LANGUAGE_BIDI': False,
 'LANGUAGE_CODE': 'en',
 'None': None,
 'SPEEDY_MATCH_DOMAIN': 'speedy.match.localhost:8020',
 'SPEEDY_NET_DOMAIN': 'speedy.net.localhost:8010',
 'True': True,
 'active_url_name': 'matches:edit_match_settings',
 'attrs': {},
 'block': <Block Node: tab_content. Contents: [<TextNode: '\n        '>, <crispy_forms.templatetags.crispy_forms_tags.CrispyFormNode object at 0x000001FA83D34278>, <TextNode: '\n    '>]>,
 'csrf_token': <SimpleLazyObject: 'Zl85ZCp494oqTL63lapgjgqmoJYciMuQiXy54hRHn3l0zumVLqOJeAzb6pStDX4H'>,
 'debug': True,
 'disable_csrf': False,
 'error_text_inline': True,
 'field_class': '',
 'flat_attrs': '',
 'form': <SpeedyMatchProfileFullMatchForm bound=True, valid=False, fields=(gender_to_match;match_description_en;min_age_to_match;max_age_to_match;diet_match;smoking_status_match;relationship_status_match)>,
 'form_action': '',
 'form_attrs': {},
 'form_class': '',
 'form_error_title': None,
 'form_id': '',
 'form_method': 'post',
 'form_show_errors': True,
 'form_show_labels': True,
 'form_style': '',
 'form_tag': True,
 'formset_error_title': None,
 'help_text_inline': False,
 'html5_required': False,
 'include_media': True,
 'inputs': [<crispy_forms.layout.Submit object at 0x000001FA83D7BE48>],
 'is_formset': False,
 'label_class': '',
 'messages': <django.contrib.messages.storage.fallback.FallbackStorage object at 0x000001FA83D69710>,
 'object': <SiteProfile: <User Profile 581018852069973 - Uri/uri-rodberg> @ Speedy Match>,
 'perms': <django.contrib.auth.context_processors.PermWrapper object at 0x000001FA83D7B828>,
 'request': '<<request>>',
 'settings': <LazySettings "speedy.match.settings.___development_uri">,
 'site': <Site: speedy.match.localhost:8020>,
 'site_name': 'Speedy Match',
 'site_title': 'Speedy Match [alpha]',
 'siteprofile': <SiteProfile: <User Profile 581018852069973 - Uri/uri-rodberg> @ Speedy Match>,
 'sites': '<<queryset of sites.Site>>',
 'sql_queries': <function debug.<locals>.<lambda> at 0x000001FA85A4FC80>,
 'template_pack': 'bootstrap4',
 'user': <SimpleLazyObject: <User: <User 581018852069973 - Uri/uri-rodberg>>>,
 'view': <speedy.match.matches.views.EditMatchSettingsView object at 0x000001FA83D28780>}
{'forloop': {'counter': 1,
             'counter0': 0,
             'first': True,
             'last': True,
             'parentloop': {},
             'revcounter': 1,
             'revcounter0': 0},
 'input': <crispy_forms.layout.Submit object at 0x000001FA83D7BE48>}
```

**Comment by carltongibson** [comment]
Right, so you've got a  

```
'form_show_errors': True,
```
in there. So looking at `bootstrap4`: 

https://github.com/django-crispy-forms/django-crispy-forms/blob/7d7fc345986bf00b7047181448af71eaec28ebcb/crispy_forms/templates/bootstrap4/uni_form.html#L5-L7

So it should be showing I think. Curious.

**Comment by uri-rodberg** [comment]
I found out we have our own `bootstrap4/render.html` template:

```
{% load crispy_forms_field %}

{% if field.is_hidden %}
    {{ field }}
{% else %}
    {% if field|is_checkbox %}
        <div class="form-group">
        {% if label_class %}
            <div class="controls col-{{ bootstrap_device_type }}-offset-{{ label_size }} {{ field_class }}">
        {% endif %}
    {% endif %}
    <{% if tag %}{{ tag }}{% else %}div{% endif %} id="div_{{ field.auto_id }}" {% if not field|is_checkbox %}class="form-group{% else %}class="checkbox{% endif %}{% if wrapper_class %} {{ wrapper_class }}{% endif %}{% if form_show_errors%}{% if field.errors %} has-error{% endif %}{% endif %}{% if field.css_classes %} {{ field.css_classes }}{% endif %}">
        {% if field.label and not field|is_checkbox and form_show_labels %}
            <label for="{{ field.id_for_label }}" class="control-label {{ label_class }}{% if field.field.required %} requiredField{% endif %}">
                {{ field.label|safe }}{% if field.field.required %}<span class="asteriskField">*</span>{% endif %}
            </label>
        {% endif %}

        {{ field }}
    </{% if tag %}{{ tag }}{% else %}div{% endif %}>
{% endif %}
```

```
bootstrap4/render.html
...\\speedy\\match\\templates\\bootstrap4\\render.html
Toggle context
{'False': False, 'None': None, 'True': True}
{'DEFAULT_MESSAGE_LEVELS': {'DEBUG': 10,
                            'ERROR': 40,
                            'INFO': 20,
                            'SUCCESS': 25,
                            'WARNING': 30},
 'False': False,
 'LANGUAGES': [('en', 'English'), ('he', 'Hebrew')],
 'LANGUAGE_BIDI': False,
 'LANGUAGE_CODE': 'en',
 'None': None,
 'SPEEDY_MATCH_DOMAIN': 'speedy.match.localhost:8020',
 'SPEEDY_NET_DOMAIN': 'speedy.net.localhost:8010',
 'True': True,
 'active_url_name': 'matches:edit_match_settings',
 'attrs': {},
 'block': <Block Node: tab_content. Contents: [<TextNode: '\n        '>, <crispy_forms.templatetags.crispy_forms_tags.CrispyFormNode object at 0x000001FA83D34278>, <TextNode: '\n    '>]>,
 'csrf_token': <SimpleLazyObject: 'Zl85ZCp494oqTL63lapgjgqmoJYciMuQiXy54hRHn3l0zumVLqOJeAzb6pStDX4H'>,
 'debug': True,
 'disable_csrf': False,
 'error_text_inline': True,
 'field': <django.forms.boundfield.BoundField object at 0x000001FA83BD7128>,
 'field_class': '',
 'flat_attrs': '',
 'form': <SpeedyMatchProfileFullMatchForm bound=True, valid=False, fields=(gender_to_match;match_description_en;min_age_to_match;max_age_to_match;diet_match;smoking_status_match;relationship_status_match)>,
 'form_action': '',
 'form_attrs': {},
 'form_class': '',
 'form_error_title': None,
 'form_id': '',
 'form_method': 'post',
 'form_show_errors': True,
 'form_show_labels': True,
 'form_style': '',
 'form_tag': True,
 'formset_error_title': None,
 'help_text_inline': False,
 'html5_required': False,
 'include_media': True,
 'inputs': [<crispy_forms.layout.Submit object at 0x000001FA83D7BE48>],
 'is_formset': False,
 'label_class': '',
 'labelclass': None,
 'messages': <django.contrib.messages.storage.fallback.FallbackStorage object at 0x000001FA83D69710>,
 'object': <SiteProfile: <User Profile 581018852069973 - Uri/uri-rodberg> @ Speedy Match>,
 'perms': <django.contrib.auth.context_processors.PermWrapper object at 0x000001FA83D7B828>,
 'request': '<<request>>',
 'settings': <LazySettings "speedy.match.settings.___development_uri">,
 'site': <Site: speedy.match.localhost:8020>,
 'site_name': 'Speedy Match',
 'site_title': 'Speedy Match [alpha]',
 'siteprofile': <SiteProfile: <User Profile 581018852069973 - Uri/uri-rodberg> @ Speedy Match>,
 'sites': '<<queryset of sites.Site>>',
 'sql_queries': <function debug.<locals>.<lambda> at 0x000001FA85A4FC80>,
 'template_pack': 'bootstrap4',
 'user': <SimpleLazyObject: <User: <User 581018852069973 - Uri/uri-rodberg>>>,
 'view': <speedy.match.matches.views.EditMatchSettingsView object at 0x000001FA83D28780>,
 'wrapper_class': None}
```

**Comment by carltongibson** [comment]
OK, so I'd have a look what else you have going on in there... (Sorry I don't think I can help you right now...)

**Comment by uri-rodberg** [comment]
I tried to add `{% include 'bootstrap4/layout/help_text_and_errors.html' %}` to our `bootstrap4/render.html` template, but it didn't work (I checked locally).

**Comment by uri-rodberg** [comment]
OK, I checked what happens if I add ` {{ form.errors }}` in the template after `{% crispy form %}`. You can see that there are 4 error messages I created on purpose. But only 2 of them are displayed in the form. See the screenshot:

![2019-08-07_1758_crispy_6](https://user-images.githubusercontent.com/12808124/62633692-3ed80980-b93d-11e9-9475-01dedbd31c0d.png)

**Comment by carltongibson** [comment]
I’m confused. I thought you were talking about `non_field_errors`... 

I think you need to eliminate all the speedy net stuff and produce a reduced example that just uses a minimal form and crispy forms basics. There’s just too much going on otherwise.

**Comment by uri-rodberg** [comment]
Yes, I thought these were non field errors but I discovered it's field errors of some of the fields, such as checkbox or radio buttons. Non field errors are displayed, as you can see in screenshot #1 above.

I will try to see if I can remove the Speedy Net stuff.

**Comment by uri-rodberg** [comment]
OK, I disabled our layout and specific template, you can see my commit and code:

https://github.com/speedy-net/speedy-net/commit/3d4464613767ed1213a0818ad86b6a36e4cdc732

https://github.com/speedy-net/speedy-net/blob/crispy_branch_2019-08-07_a/speedy/match/matches/forms.py#L16-L31

But the error messages still don't display on fields such as radio buttons:

![2019-08-07_1843_crispy_7](https://user-images.githubusercontent.com/12808124/62637120-89f51b00-b943-11e9-9f62-81c275b23106.png)

**Comment by uri-rodberg** [comment]
I checked now and it works with django-crispy-forms==1.6.1! I think we will just downgrade to django-crispy-forms==1.6.1.

Screenshot:
![2019-08-07_1902_crispy_8](https://user-images.githubusercontent.com/12808124/62638542-5ff12800-b946-11e9-96a2-2afeb4ee81f3.png)

**Comment by uri-rodberg** [comment]
The problem is only with crispy forms versions 1.7 and above. So I think it's a bug of Django crispy forms.

**Comment by uri-rodberg** [comment]
I can see that the commit that caused this error is 34ebbd74d5b6c209f82e6f6c3d4d05bd3a48e18a (1.7.0). The commit before works. But this commit changed many lines, so I don't know which one cause this issue.

**Comment by uri-rodberg** [comment]
OK I found it, commit b20fb75496ea3e96e412234c25aec3de31e70330 is the first commit not working. The previous commit 3b0de89d9d490f7e902c9a879b64021e3861c143 is working.

**Comment by uri-rodberg** [comment]
OK, I found the problem - we had a display none CSS when not needed. It's not a bug in Django Crispy Forms, I'm sorry about the confusion. If was a bug in our CSS code. I fixed it with commit ad1c9ba9ef23529c72e62ed31f969af19f6b633a. Thanks for your time! I'm closing this issue.

**Comment by carltongibson** [comment]
Super. Thanks for the follow-up. Maybe it helps someone else in the future!
